src/Container/ApplicationFactory.php

```diff
@@ -165,9 +166,20 @@ private function injectRoutes(Application $app, ContainerInterface $container)
             $methods = (isset($spec['allowed_methods']) && is_array($spec['allowed_methods']))
                 ? $spec['allowed_methods']
                 : null;
-            $name    = isset($spec['name']) ? $spec['name'] : null;
+            $name = isset($spec['name']) ? $spec['name'] : null;
             $methods = (null === $methods) ? Route::HTTP_METHOD_ANY : $methods;
-            $route   = new Route($spec['path'], $spec['middleware'], $methods, $name);
+
+            if (!is_callable($spec['middleware']) && is_array($spec['middleware'])) {
+                $middlewarePipe = new MiddlewarePipe();
+
+                foreach ($spec['middleware'] as $middleware) {
+                    $middlewarePipe->pipe(is_callable($middleware) ? $middleware : $container->get($middleware));
+                }
+
+                $spec['middleware'] = $middlewarePipe;
+            }
```

Comment:
@weierophinney I noticed that the changes thus far would not (yet) allow this type of array configuration in the pre- and post-routing pipelines. Of course, one could just specify them separately, like this:

```php
return [
    'middleware_pipeline' => [
        'pre_routing' => [
            [
                'middleware' => 'Foo',
                'path' => '/foobar',
            ],
            [
                'middleware' => 'Bar',
                'path' => '/foobar',
            ],
            [
                'middleware' => 'Foobar',
                'path' => '/foobar',
            ],
        ],
    ],
];
```

but something like this would be much more concise and potentially easier to read:

```php
return [
    'middleware_pipeline' => [
        'pre_routing' => [
            [
                'middleware' => [
                    'Foo',
                    'Bar',
                    'FooBar'
                ],
                'path' => '/foobar',
            ],
        ],
    ],
];
```

Do you agree? Would you want to try to cram that into this PR as well or do it as a future PR hopefully still targeting 1.0.0?